main/lsp/requests/completion.cc

```diff
@@ -92,36 +94,94 @@ string methodSnippet(const core::GlobalState &gs, core::SymbolRef method) {
     return fmt::format("{}({}){}", shortName, fmt::join(typeAndArgNames, ", "), "${0}");
 }
 
+/**
+ * Retrieves the documentation above a symbol.
+ * - Assumes that valid ruby syntax is used.
+ * - Strips the first whitespace character from a comment e.g
+ *      # a comment
+ *      #a comment
+ *   are the same.
+ */
 optional<string> findDocumentation(string_view sourceCode, int beginIndex) {
-    // everything in the file before the method definition.
-    auto preDefinition = sourceCode.substr(0, sourceCode.rfind('\n', beginIndex));
+    // Everything in the file before the method definition.
+    string_view preDefinition = sourceCode.substr(0, sourceCode.rfind('\n', beginIndex));
 
-    int lastNewlineLoc = preDefinition.rfind('\n');
-    // if there is no '\n' in preDefinition, we're at the top of the file.
-    if (lastNewlineLoc == preDefinition.npos) {
-        return nullopt;
-    }
-    auto prevLine = preDefinition.substr(lastNewlineLoc, preDefinition.size() - lastNewlineLoc);
-    if (prevLine.find('#') == prevLine.npos) {
+    // Get all the lines before it.
+    std::vector<string_view> all_lines = absl::StrSplit(preDefinition, '\n');
+
+    // if there are no lines before the method definition, we're at the top of the file.
+    if (all_lines.empty()) {
         return nullopt;
     }
 
-    string documentation = "";
-    // keep looking for previous newline locations, searching for lines with # in them.
-    while (prevLine.find('#') != prevLine.npos) {
-        documentation = absl::StrCat(prevLine.substr(prevLine.find('#') + 1, prevLine.size()), "\n", documentation);
-        int prevNewlineLoc = preDefinition.rfind('\n', lastNewlineLoc - 1);
-        // if there is no '\n', we're at the top of the file, so just return documentation.
-        if (prevNewlineLoc == preDefinition.npos) {
+    std::vector<string_view> documentation_lines;
+
+    // Iterate from the last line, to the first line
+    for (auto it = all_lines.rbegin(); it != all_lines.rend(); it++) {
+        string_view line = absl::StripAsciiWhitespace(*it);
+
+        // Short circuit when line is empty
+        if (line.empty()) {
+            break;
+        }
+
+        // Handle single-line sig block
+        else if (absl::StartsWith(line, "sig")) {
+            // Do nothing for a one-line sig block
+        }
+
+        // Handle multi-line sig block
+        else if (absl::StartsWith(line, "end")) {
+            // ASSUMPTION: We either hit the start of file, a `sig do` or an `end`
+            do
+                it++;
```

Comment:
This loop never updates `line`, so it will never find `sig do` or `end`.

Reply:
An alternative way to perform this parsing with only a single loop (and no inner loop) is to have state variables that the loop updates.

Example: `bool is_in_sig_block` outside of the loop. The first encounter of `end` would set it to `true`. If you encounter `end` again, we break the loop.

I believe this would also reduce the amount of code involved, since you can repurpose the existing `StartsWith(line, "sig")` conditional.

Reply:
Whoops -- luckily the test in lsp_test catches this.

I actually tried an approach earlier, but it required me to check that variable in other parts of the loop. It got a bit messy (e.g sig with is_in_sig_block is different from sig and ! is_in_sig_block)so I thought the inner loop makes each if block independent of what was seen before.

Reply:
Refreshed the line in 1d7502718, test passes now